File: connectwave-backend/src/main/java/com/abdulmo123/connectwave/service/FriendshipService.java
package com.abdulmo123.connectwave.service;

import com.abdulmo123.connectwave.dto.FriendshipDto;
import com.abdulmo123.connectwave.entity.Friendship;

import java.util.List;

public interface FriendshipService {

    FriendshipDto sendFriendshipRequest(Long senderId, Long receiverId);

    Friendship existingFriendshipRequest(Long senderId, Long receiverId);

    void cancelSentFriendshipRequest(Long senderId, Long receiverId);

    Friendship respondToFriendshipRequest(Long senderId, Long receiverId, String action);

    void removeExistingFriendship(Long senderId, Long receiverId);

    FriendshipDto getNewFriendshipRequest(Long senderId, Long receiverId);

}

File: connectwave-frontend/src/app/services/friendship.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment.development';
import { User } from '../models/user';
import { Friendship } from '../models/friendship';

@Injectable({
  providedIn: 'root'
})
export class FriendshipService {

  constructor(private http: HttpClient) { }

  public addFriend(senderId: number, receiverId: number): Observable<Friendship> {
    return this.http.post<Friendship>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/sendFriendshipRequest`,
      {});
  }

  public getExistingFriendshipRequest(senderId: number, receiverId: number): Observable<Friendship> {
    return this.http.get<Friendship>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/existingFriendshipRequest`);
  }

  public cancelSentFriendshipRequest(senderId: number, receiverId: number): Observable<void> {
    return this.http.delete<void>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/cancelSentFriendshipRequest`);
  }

  public responseToFriendshipRequest(senderId: number, receiverId: number, action: string): Observable<Friendship> {
    return this.http.post<Friendship>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/${action}/respondToFriendshipRequest`,
      {})
  }

  public getExistingFriendshipRelationship(senderId: number, receiverId: number): Observable<Friendship> {
    return this.http.get<Friendship>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/existingFriendshipRelationship`);
  }

  public removeExistingFriend(senderId: number, receiverId: number): Observable<void> {
    return this.http.delete<void>(
      `${environment.hostUrl}/api/v1/friendships/${senderId}/${receiverId}/removeExistingFriend`);
  }
}
